r55/src/exec.rs

```diff
@@ -133,32 +136,89 @@ pub fn handle_register<EXT, DB: Database>(handler: &mut EvmHandler<'_, EXT, DB>)
     // execute riscv context or old logic.
     let old_handle = handler.execution.execute_frame.clone();
     handler.execution.execute_frame = Arc::new(move |frame, memory, instraction_table, ctx| {
-        let result = if let Some(Some(riscv_context)) = call_stack.borrow_mut().first_mut() {
+        let depth = call_stack.borrow().len() - 1;
+
+        // use last frame as stack is LIFO
+        let mut result = if let Some(Some(riscv_context)) = call_stack.borrow_mut().last_mut() {
+            debug!(
+                "=== [FRAME-{}] Contract: {} ============-",
+                depth,
+                frame.interpreter().contract.target_address,
+            );
             execute_riscv(riscv_context, frame.interpreter_mut(), memory, ctx)?
         } else {
+            debug!("=== [OLD Handler] ==================--");
             old_handle(frame, memory, instraction_table, ctx)?
         };
 
-        // if it is return pop the stack.
+        // if action is return, pop the stack.
         if result.is_return() {
             call_stack.borrow_mut().pop();
+
+            let is_last = call_stack.borrow().is_empty();
+            debug!(
+                "=== [FRAME-{}] Ouput: RETURN {}",
+                depth,
+                if is_last { "(last)" } else { "" }
+            );
+
+            if !is_last {
+                if let Some(Some(parent)) = call_stack.borrow_mut().last_mut() {
+                    if let Some(return_range) = &parent.returned_data_destiny {
+                        if let InterpreterAction::Return { result: res } = &mut result {
+                            // get allocated memory slice
+                            let return_memory = parent
+                                .emu
+                                .cpu
+                                .bus
+                                .get_dram_slice(return_range.clone())
+                                .expect("unable to get memory from return range");
+
+                            debug!("- Return data: {:?}", res.output);
+                            debug!("- Memory range: {:?}", return_range);
+                            debug!("- Memory size: {}", return_memory.len());
+
+                            // write return data to parent's memory
```

Comment:
This only supports the case where a RISCV contract calls a RISCV contract, right?

Reply:
yeah, i totally missed that